lib/ex_unit/lib/ex_unit/filters.ex

```diff
@@ -14,16 +14,31 @@ defmodule ExUnit.Filters do
   on the command line) includes a line number filter, and if so returns the
   appropriate ExUnit configuration options.
   """
-  @spec parse_path(String.t()) :: {String.t(), any}
+  @spec parse_path(String.t()) :: {String.t(), Keyword.t()}
   def parse_path(file) do
-    {paths, [line]} = file |> String.split(":") |> Enum.split(-1)
+    {paths, lines, _, _} =
+      file
+      |> String.split(":")
+      |> Enum.reduce({[], [], false, false}, &reduce_line_numbers/2)
 
-    case Integer.parse(line) do
-      {_, ""} ->
-        {Enum.join(paths, ":"), exclude: [:test], include: [line: line]}
+    path = Enum.join(paths, ":")
 
-      _ ->
-        {file, []}
+    case lines do
+      [] -> {path, []}
+      _ -> {path, exclude: [:test], include: Enum.map(lines, &{:line, &1})}
+    end
+  end
+
+  defp reduce_line_numbers(part, {paths, _, false, false}) do
+    {paths ++ [part], [], false, String.match?(part, ~r/\.exs/)}
```

Comment:
I would prefer to avoid using regexes. Here is a general algorithm I would propose:

1. Split the path into ":"
2. Revert the split path and call Enum.split_while so you take trailing information while `match?({_, ""}, Integer.parse(part))` returns true
3. Once that is done, reverse the path back, join the remaining parts with `:` and compute the line numbers

This should be cleaner and rely less on stateful reduce.

Thoughts?

Reply:
To be more precise, something like this:

```elixir
path
|> String.split(path, ":")
|> extract_line_numbers()
|> case do
  {[], _} ->
    {path, []}

  {reverse_lines, reverse_parts} ->
    include = reverse_lines |> Enum.reverse() |> Enum.map(&{:line, String.to_integer(&1)})
    path = reverse_parts |> Enum.reverse() |> Enum.join(":")
    {path, exclude: [:test], include: include}
end

defp extract_line_numbers([part]), do: {[], [part]}

defp extract_line_numbers(parts) do
  parts
  |> Enum.reverse()
  |> Enum.split_while(&match?({_, ""}, Integer.parse(&1)))
end
```

Reply:
What about? (a little refactor) :)

```elixir
case extract_line_numbers(path) do
  {path, []} -> {path, []}
  {path, line_numbers} -> {path, exclude: [:test], include: line_numbers}
end

defp extract_line_numbers(path) do
  case String.split(path, ":") do
    [part] -> 
      {path, []}

    parts ->
      {reversed_line_numbers, reversed_path_parts} =
        parts
        |> Enum.reverse()
        |> Enum.split_while(&match?({_, ""}, Integer.parse(&1)))

      line_numbers = 
        reversed_line_numbers 
        |> Enum.reverse() 
        |> Enum.map(&{:line, String.to_integer(&1)})
        
      path = 
        reversed_path_parts 
        |> Enum.reverse() 
        |> Enum.join(":")

      {path, line_numbers}
  end
end
```

Reply:
I did consider something like that, but that would change the semantics of the feature. If we were do to that, the following test would change from:

```elixir
    assert ExUnit.Filters.parse_path("test/some/path.exs:123notalinenumber123:456") ==
             {"test/some/path.exs:123notalinenumber123:456", []}
```

to

```elixir
    assert ExUnit.Filters.parse_path("test/some/path.exs:123notalinenumber123:456") ==
             {"test/some/path.exs:123notalinenumber123", [exclude: [:test], include: [line: "456"]]}
```

That would create some weird behavior since then the following test (with those same line numbers but in different order) would remain the same as:

```elixir
    assert ExUnit.Filters.parse_path("test/some/path.exs:123:456notalinenumber456") ==
             {"test/some/path.exs:123:456notalinenumber456", []}
```